### DISCUSSION TITLE
Avoid duplication in project.json

### DISCUSSION THREAD

**Discussion by BeyondEvil** [nrwl/nx#9745] [Feature Requests]
We have a NX workspace that currently hosts five applications (and growing).

In each `project.json` I have (among other things obv.) this:

```json
    "docker-build": {
      "executor": "@nrwl/workspace:run-commands",
      "options": {
        "commands": [
          "docker pull \"${ECR_REGISTRY}/co/platform-api:build\"",
          "docker build -t \"${ECR_REGISTRY}/co/platform-api:build\" -f docker/ts/Dockerfile --cache-from \"${ECR_REGISTRY}/co/platform-api:build\" --target build --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg APP='platform-api' .",
          "docker build -t platform-api:latest -f docker/ts/Dockerfile --cache-from \"${ECR_REGISTRY}/co/platform-api:build\" --target production --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg APP='platform-api' .",
          "docker push \"${ECR_REGISTRY}/co/platform-api:build\""
        ],
        "parallel": false
      }
    },
    "docker-push": {
      "executor": "@nrwl/workspace:run-commands",
      "options": {
        "commands": [
          "docker tag platform-api:latest \"${ECR_REGISTRY}/co/platform-api:$(git-slug-for-app.sh platform-api)\"",
          "docker push \"${ECR_REGISTRY}/co/platform-api:$(git-slug-for-app.sh platform-api)\""
        ],
        "parallel": false
      }
    },
```

Where the _only_ difference between the apps is the app name (`platform-api`) in the example above.

Please tell me I'm doing it wrong, and that there's a way of defining a target _once_ and then passing in the app name?

Also, is there a way of _not_ having to repeat oneself inside `commands`? They seem to each run in their own fork, which means I can't do:

```json
    "docker-push": {
      "executor": "@nrwl/workspace:run-commands",
      "options": {
        "commands": [
          "repo_url=\"${ECR_REGISTRY}/co/platform-api:$(git-slug-for-app.sh platform-api)\""
          "docker tag platform-api:latest \"${repo_url}\"",
          "docker push  \"${repo_url}\""
        ],
        "parallel": false
      }
    },
```

🙏

**Comment by gioragutt** [comment]
You can use project inference (docs is WIP, you can find a draft [here](https://nx-dev-git-fork-agentender-docs-nx-pluginproject-inference-nrwl.vercel.app/extending-nx/project-inference-plugins)).

Here's an example of something you could do, following the example from the docs:

```ts
import { TargetConfiguration } from '@nrwl/devkit';
import path from 'path';

export const projectFilePatterns = ['Dockerfile'];

export function registerProjectTargets(
  projectFilePath: string,
): Record<string, TargetConfiguration> {
  const appName = path.dirname(projectFilePath);

  return {
    'docker-build': {
      executor: '@nrwl/workspace:run-commands',
      options: {
        commands: [
          `docker pull "\${ECR_REGISTRY}/co/${appName}:build"`,
          `docker build -t "\${ECR_REGISTRY}/co/${appName}:build" -f docker/ts/Dockerfile --cache-from "\${ECR_REGISTRY}/co/${appName}:build" --target build --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg APP=\'${appName}\' .`,
          `docker build -t ${appName}:latest -f docker/ts/Dockerfile --cache-from "\${ECR_REGISTRY}/co/${appName}:build" --target production --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg APP=\'${appName}\' .`,
          `docker push "\${ECR_REGISTRY}/co/${appName}:build"`,
        ],
        parallel: false,
      },
    },
    'docker-push': {
      executor: '@nrwl/workspace:run-commands',
      options: {
        commands: [
          `docker tag ${appName}:latest "\${ECR_REGISTRY}/co/${appName}:$(git-slug-for-app.sh ${appName})"`,
          `docker push "\${ECR_REGISTRY}/co/${appName}:$(git-slug-for-app.sh ${appName})"`,
        ],
        parallel: false,
      },
    },
  };
}
```

You can do it without a full plugin library and just use a `.js` file (like you can see [here](https://github.com/nrwl/nx/pull/9697/files#diff-f7e0cfcb2e9289565b3cb0fb0997c501f91901898d1a38fb53b93388a5978e8e), and referenced [here](https://github.com/nrwl/nx/pull/9697/files#diff-15552e50b1b7c05b05b7ffe08ee47b5ed62b8d2039229c58972a42fd22a7381c)).

Feel free to try it out and let me know if this works. This is new territory for me as well, I'm curious as to whether this works as expected or not. Good luck!

**Reply by BeyondEvil** [reply]
I'll give it a go and report back, thanks!

**Reply by BeyondEvil** [reply]
According to this from the docs:

> If workspace.json was present, my-project wouldn't be able to be inferred, since full project inference requires workspace.json to be absent.

it doesn't work if `workspace.json` is present. We have both `nx.json` and `workspace.json`.

Does that mean we can't use this? Or are we doing something wrong? 🤔

**Reply by gioragutt** [reply]
Probably doing something wrong because in the example I linked to in the PR, it is used to add a target to all packages, and it's the NX repo so it has a `workspace.json`.

I'll try to look at the source code of this feature today to understand the specifics

**Reply by gioragutt** [reply]
cc @AgentEnder

**Reply by AgentEnder** [reply]
"workspace.json" disables **_project_** inference, but **_target_** inference is still active.

**Reply by gioragutt** [reply]
@AgentEnder perhaps this is a good thing to specify in the docs, to make this distinction clear

**Reply by Wscats** [reply]
This feature https://github.com/nrwl/nx/pull/9441 can solve this problem, platform-api can be configured as a variable in project.json for reuse. @gioragutt @AgentEnder

**Reply by BeyondEvil** [reply]
> This feature #9441 can solve this problem, platform-api can be configured as a variable in project.json for reuse. @gioragutt @AgentEnder

That’s only part of the issue.

The bigger issue we’re trying to solve is to not have to duplicate the targets in _every_ apps project.json.

**Reply by gioragutt** [reply]
I sometimes compare NX to gradle.

When you enable a plug-in (which is something that can even be in the project level), you get a bunch of tasks (targets,executors).

This is a behavior we can compare project/target inference to.

**Comment by BeyondEvil** [comment]
Finally had some time to look at this.

In `tools/scripts/docker-build.js` I have:

<details>

<summary>docker-build.js</summary>

```js
const path = require('path');

module.exports = {
  projectFilePatterns: ['Dockerfile'],
  registerProjectTargets: (f) => {
    const appName = path.dirname(f)
    const ecrRegistry = process.env.ECR_REGISTRY
    return {
    'docker-build': {
      executor: '@nrwl/workspace:run-commands',
      options: {
        commands: [
          `docker pull "${ecrRegistry}/co/${appName}:build" || true`,
          `docker build -t "${ecrRegistry}/co/${appName}:build" -f docker/ts/Dockerfile --cache-from "${ecrRegistry}/co/${appName}:build" --target build --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg APP=${appName} .`,
          `docker build -t ${appName}:latest -f docker/ts/Dockerfile --cache-from "${ecrRegistry}/co/${appName}:build" --target production --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg APP=${appName} .`,
          `docker push "${ecrRegistry}/co/${appName}:build"`,
        ],
        parallel: false,
      },
    },
  }},
};
```

</details>

<details>

<summary>nx.json</summary>

```json
{
  "npmScope": "px-platform",
  "affected": {
    "defaultBase": "master"
  },
  "workspaceLayout": {
    "appsDir": "apps",
    "libsDir": "packages"
  },
  "cli": {
    "defaultCollection": "@nrwl/nest"
  },
  "implicitDependencies": {
    "package.json": {
      "dependencies": "*",
      "devDependencies": "*"
    },
    ".eslintrc.json": "*"
  },
  "tasksRunnerOptions": {
    "default": {
      "runner": "@nrwl/workspace/tasks-runners/default",
      "options": {
        "cacheableOperations": ["build", "lint", "test"]
      }
    }
  },
  "targetDependencies": {
    "build": [
      {
        "target": "build",
        "projects": "dependencies"
      }
    ],
    "docker-build": [
      {
        "target": "docker-build",
        "projects": "self"
      }
    ]
  },
  "plugins": ["./tools/scripts/docker-build.js"],
  "defaultProject": "platform-api"
}
```

</details>

When I try to run:

```sh
$ npx nx run dqs:docker-build
```

I get:

```sh
>  NX   Cannot find target 'docker-build' for project 'dqs'
```

<details>

<summary>workspace.json</summary>

```json
{
  "version": 2,
  "projects": {
    "dpop-jwt": "packages/dpop-jwt",
    "dqs": "apps/dqs/app",
    "ins": "apps/ins/app",
    "ivs": "apps/ivs/app",
    "notifications-api": "apps/notifications-api/app",
    "platform-api": "apps/platform-api/app",
    "platform-libs": "packages/platform-libs",
    "remote-config": "packages/remote-config",
    "testing": "packages/testing"
  }
}
```

</details>

```sh
$ npx nx --version
13.8.4
```

What am I missing/doing wrong? @gioragutt

**Reply by gioragutt** [reply]
I added a similar plugin today in a repo, this is just what's warned about in the last section in the docs that @AgentEnder is writing.

Try running `nx daemon` to make it re-evaluate the plugins and tasks of each project

**Reply by BeyondEvil** [reply]
Ah, RTFM, sorry about that 😂 

That moved me forward, but now I hit this:

```sh
>  NX   Could not execute dqs:docker-build because it has a circular dependency

   dqs:docker-build --> dqs:docker-build
```

Also, what is `projectFilePatterns` used for? What should I set it to in this instance? @gioragutt

**Reply by BeyondEvil** [reply]
I should mention that I set `projectFilePatterns` to `['project.json']`.

If I had it as `['Dockerfile']` I still got the `Cannot find target...` even after running `npx nx daemon`.

**Reply by gioragutt** [reply]
`dqs:docker-build --> dqs:docker-build` means that in your `nx.json` you're making `docker-build` depend on `docker-build` of `self`. At least that's what I think.

`projectFilePatterns` is used to filter files to infer projects. IMO the API is a bit lackluster but it is what it is for now.

So I think the way it behaves now is that it looks for projects that have a `Dockerfile`, and adds it to them.
IDK how it handles nested files (f.e: `apps/my-app/configs/Dockerfile`), that's something I intend to find out soon.

**Reply by BeyondEvil** [reply]
Ah, I see. Ok, I changed from `self` to `dependencies`, and that removed the circular error.

However, now it's trying to build dependencies, which doesn't have a Dockerfile (which none of the projects have). We have a Dockerfile-template that is generated as part of the `docker-build` target (`-f docker/ts/Dockerfile`). 😅

**Reply by BeyondEvil** [reply]
Essentially what I would want is for it to only run against `apps` and not `packages`:

```json
  "workspaceLayout": {
    "appsDir": "apps",
    "libsDir": "packages"
  },
```

**Reply by BeyondEvil** [reply]
Something like this could work (but doesn't, unless I'm doing something wrong):

```js
projectFilePatterns: ['apps/**/project.json'],
```

**Reply by BeyondEvil** [reply]
According to this, it should work:

> Plugins can provide an array of glob patterns, projectFilePatterns that are used to infer project information.

https://github.com/nrwl/nx/blob/f3dde188324004dee2bba00de2a090d1b4c010d7/docs/shared/devkit-and-nx-plugins.md

🤔

**Reply by gioragutt** [reply]
So? Is it working with the glob?

**Reply by BeyondEvil** [reply]
No, it doesn't seem like it is. Looking at the code [here](https://github.com/nrwl/nx/blob/8af847c579543b8c6708ecdd850f5d02240b874f/packages/nx/src/utils/nx-plugin.ts#L90-L91) it seems it starts the search at the project root, instead of workspace root. @gioragutt

**Reply by gioragutt** [reply]
So doesn't just `"project.json"` work?

**Reply by BeyondEvil** [reply]
Yes, but that also picks up our "local" dependencies in our `packages` directory.

```json
  "workspaceLayout": {
    "appsDir": "apps",
    "libsDir": "packages"
  },
```

And we don't want that, since they're never containerized. @gioragutt

**Reply by gioragutt** [reply]
How about filtering `filename.includes('/apps/')` in the method?

**Reply by BeyondEvil** [reply]
Like this:

```js
registerProjectTargets: (f) => {
    const appName = path.dirname(f.includes('/apps/'))
```

Cause it didn't work 😬

**Reply by gioragutt** [reply]
1. is `f` a relative or absolute path?
2. did you just `return;` (or `return {};`) after it?

**Reply by BeyondEvil** [reply]
If I log out `f` it _only_ contains the dependencies:

```
packages/dpop-jwt
packages/testing
packages/dpop-jwt
```

That's relative to the workspace root.

Using

```js
  projectFilePatterns: ['project.json'],
  registerProjectTargets: (f) => {
    const appName = path.dirname(f)
    if ( f.includes('apps') ) return {}
    const ecrRegistry = process.env.ECR_REGISTRY
    return {
```

I get the ` Cannot find target 'docker-build' for project 'dqs'` error again.

Feels like going in circles 😂  @gioragutt

**Reply by BeyondEvil** [reply]
Nvm, had forgotten to negate it! 😂 

**coding while tired**

**Reply by gioragutt** [reply]
lol 😅

so you good then?

**Reply by BeyondEvil** [reply]
Sorry for the late reply, me and Node does _not_ get along. 😅 

This is the "finished" targets:

```js
const path = require('path');
fs = require('fs');

module.exports = {
  projectFilePatterns: ['project.json'],
  registerProjectTargets: (f) => {
    // packages are not containerized
    if (! f.includes('apps') ) return {};

    const appName = path.dirname(f).match(/apps\/(\S+)\/app/)[1]
    const ecrRegistry = process.env.ECR_REGISTRY
    const image = `${ecrRegistry}/proxyco/${appName}`

    return {
    'docker-build': {
      executor: '@nrwl/workspace:run-commands',
      options: {
        commands: [
          `docker pull ${image}:build || true`,
          `docker build -t ${image}:build -f docker/ts/Dockerfile --cache-from ${image}:build --target build --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg APP=${appName} .`,
          `docker build -t ${appName}:latest -f docker/ts/Dockerfile --cache-from ${image}:build --target production --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg APP=${appName} .`,
          `docker push ${image}:build`,
        ],
        parallel: false,
      },
    },
    'docker-push': {
      executor: '@nrwl/workspace:run-commands',
      options: {
        commands: [
          `docker tag dqs:latest ${image}:$(get-app-hash.sh ${appName})`,
          `docker push ${image}:$(get-app-hash.sh ${appName})`
        ],
        parallel: false,
      },
    },
  }},
};
```

Thanks for your help and patience! 🙇

**Reply by gioragutt** [reply]
```ts
// @ts-check

const path = require('path');

/**
 * @type {import('@nrwl/devkit').NxPlugin}
 */
module.exports = {
  projectFilePatterns: ['project.json'],
  registerProjectTargets: file => {
    // packages are not containerized
    if (!file.includes('apps')) return {};

    const appName = path.dirname(file).match(/apps\/(\w+)\/app/)?.[1];

    if (!appName) return {};

    const ecrRegistry = process.env.ECR_REGISTRY;
    const image = `${ecrRegistry}/co/${appName}`;

    return {
      'docker-build': {
        executor: '@nrwl/workspace:run-commands',
        options: {
          commands: [
            `docker pull ${image}:build || true`,
            `docker build -t ${image}:build -f docker/ts/Dockerfile --cache-from ${image}:build --target build --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg APP=${appName} .`,
            `docker build -t ${appName}:latest -f docker/ts/Dockerfile --cache-from ${image}:build --target production --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg APP=${appName} .`,
            `docker push ${image}:build`,
          ],
          parallel: false,
        },
      },
      'docker-push': {
        executor: '@nrwl/workspace:run-commands',
        options: {
          commands: [
            `docker tag dqs:latest ${image}:$(get-app-hash.sh ${appName})`,
            `docker push ${image}:$(get-app-hash.sh ${appName})`,
          ],
          parallel: false,
        },
      },
    };
  },
};
```

Try this, I cleaned it up a little for you. The `// @ts-check` and `@type {...}` comments help with type checking in `.js` files, so you get a little more completion and validation.

**Reply by gioragutt** [reply]
@BeyondEvil Good job, glad to see you got this through 👏🏻

**Reply by BeyondEvil** [reply]
> @BeyondEvil Good job, glad to see you got this through 👏🏻

Thank you, appreciate it! @gioragutt 

You managed to answer before I edited my comment. I managed to figure out what the issue was, my regex was incorrect (`\w+`), I needed (`\S+`) since we have apps with a dash in the name. 😅

**Reply by gioragutt** [reply]
Oh awesome, perhaps it'll help me in the future when I need to make such a regex myself :)
